sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -636,6 +641,96 @@ object GpuCast extends Arm {
+                                 legacyCastToString:          Boolean,
+                                 stringToDateAnsiModeEnabled: Boolean): ColumnVector = {
+    val (leftStr, rightStr) =  ("[", "]")
+    val emptyStr = ""
+    val spaceStr = " "
+    val nullStr = if (legacyCastToString) emptyStr  else "null"
+    val sepStr = "," + (if (legacyCastToString) emptyStr else spaceStr)
+    val numRows = input.getRowCount.toInt
+
+    withResource(
+      Seq(leftStr, rightStr, emptyStr, spaceStr, nullStr, sepStr).safeMap(Scalar.fromString)
+    ){ case Seq(left, right, empty, space, nullRep, sep) =>
+
+      /* -------------------------------- helper functions -----------------------*/
+      def castChildToStr(child: ColumnView): ColumnVector = {
+        // cast each element to string
+        withResource(
+          doCast(child, elementType, StringType, ansiMode,
+                 legacyCastToString, stringToDateAnsiModeEnabled)
+        ) { strChildWithNull =>
+          withResource(strChildWithNull.replaceNulls(nullRep)) { strChild =>
+            if (legacyCastToString) {// add a space string to each non-null element
+              withResource(ColumnVector.fromScalar(space, child.getRowCount.toInt)) { spaceVec =>
+                withResource(
+                  ColumnVector.stringConcatenate(Array(spaceVec, strChild))
+                ) { addSpace =>
+                  withResource(child.isNotNull) {_.ifElse(addSpace, strChild)}
+                }
+              }
+            }
+            else { strChild.incRefCount }
+          }
+        }
+      }
+
+      /**
+       * If the first char of a string is ' ', remove it
+       * @param strVec a string type column vector
+       */
+      def removeFirstSpace(strVec: ColumnVector): ColumnVector = {
+        if (legacyCastToString){
+          withResource(strVec.substring(0,1)) { fstChar =>
+            withResource(strVec.substring(1)) { remain =>
+              withResource(fstChar.equalTo(space)) {_.ifElse(remain, strVec)}
+            }
+          }
+        }
+        else {strVec.incRefCount}
+      }
+
+      /**
+       * Add brackets to each string. Ex: ["1, 2, 3", "4, 5"] => ["[1, 2, 3]", "[4, 5]"]
+       * @param strVec a string vector
+       */
+      def addBrackets(strVec: ColumnVector): ColumnVector = {
+        withResource(
+          Seq(left, right).safeMap(s => ColumnVector.fromScalar(s, numRows))
+        ) { case Seq(leftColumn, rightColumn) =>
+          ColumnVector.stringConcatenate(empty, nullRep, Array(leftColumn, strVec, rightColumn))
+        }
+      }
+      /* -------------------------------- helper functions -----------------------*/
+
+
+      // cast child column to string type
+      withResource(input.getChildColumnView(0)) { childView =>
+        withResource(castChildToStr(childView)){ stringChild =>
+          withResource(input.replaceListChild(stringChild)) {pureStrArr =>
```

Comment:
Thanks a lot for your advice, Bobby! I strongly agree with you. The "complicated" solution is designed for some corner cases. I should add more document.
 
When `legacyCastToString = true` and there are `null` elements in the array, things can be somehow complicated. `null` elements are represented by empty strings and separated by `","`,  while other elements are separated by `", "`.  The same rule is applied on `castStructToString`:
https://github.com/NVIDIA/spark-rapids/blob/877c1b275bc8aa3df49386c882355f089f08ad17/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala#L669-L679
However, unlike CPU-version Spark or `CastStructToString` in Rapids, we could not process each element one-by-one in array, because it is implemented as `list_column` in CUDF.  As a result, we use a 6-step method to achieve it:

step-1. Cast all `not-null` elements in the array to string type
step-2. Add `space char` in the front of all `not-null` elements
step-3. Cast all `null` elements to empty strings
step-4. Concatenate elements in the array, separated by `","`
step-5. If there is a `space char` at the head of the string, remove it
step-6. Add brackets.

Here is an example:
Given an integer array: `[0, 1, 2, null, 3]` and enable `legacyCastToString`
step-1 => `["0", "1", "2", null, "3"]`
step-2 => `[" 0", " 1", " 2", null, " 3"]`
step-3 => `[" 0", " 1", " 2", "", " 3"]`
step-4 => `" 0, 1, 2,, 3"`
step-5 => `"0, 1, 2,, 3"`
step-6 => `"[0, 1, 2,, 3]"`

Reply:
Passing a `nullRep` parameter to the `stringConcatenateListElement` function is also weird. I agree with you. There are three different `stringCOncatenateListElement` APIs provided by Cudf:
The first one, which is simplist:
https://github.com/rapidsai/cudf/blob/83adc4b6cf2e1081ae77c28d421c60cc2d736165/java/src/main/java/ai/rapids/cudf/ColumnView.java#L2422-L2428
We can use it, but here is a problem. This function returns `null` for empty list, but what we want is an empty string. In this way, we might have to add another step to convert `null` to empty string, such as
```
withResource(arrayOfStrings.stirngConcatenateListElements(Scalar.fromStirng(",")){ stringColContainNulls =>
  withResource(stirngCol.replaceNulls(Scalar.fromString(""))){ stringCol =>
    ... 
  }
}
```
I am not sure whether it is more readable.

The 2nd and 3rd API both have `nullRep` to be passed
https://github.com/rapidsai/cudf/blob/83adc4b6cf2e1081ae77c28d421c60cc2d736165/java/src/main/java/ai/rapids/cudf/ColumnView.java#L2449-L2451
https://github.com/rapidsai/cudf/blob/83adc4b6cf2e1081ae77c28d421c60cc2d736165/java/src/main/java/ai/rapids/cudf/ColumnView.java#L2479-L2480